### PR TITLE
feat: sync contact on creation

### PR DESCRIPTION
**Pull request by tonyco97**
**Reference**:
- https://github.com/NethServer/dev/issues/7129

### CODE REVIEW

File: store/phonebook.go [hunk 1]
```diff
@@ -424,6 +424,67 @@ func DeletePhonebookEntryByID(ctx context.Context, id int64) error {
...
+	// Always release the lock, even on error.
+	defer func() {
+		if _, unlockErr := conn.ExecContext(ctx, "UNLOCK TABLES"); unlockErr != nil {
+			logs.Log("[ERROR][PHONEBOOK] Failed to unlock tables after centralized sync: " + unlockErr.Error())
+		}
+	}()
+
+	if _, err := conn.ExecContext(ctx, "DELETE FROM `phonebook`.`phonebook` WHERE sid_imported = 'nethcti'"); err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (tonyco97) [reply store/phonebook.go:488]: The literal "stage in a separate table and swap/replace the whole table" approach isn't applicable here: `phonebook.phonebook` is shared — it also holds rows from external sync sources (`sid_imported != 'nethcti'`: MSSQL/CSV/Infinity). Swapping the whole table would drop those contacts, which is a worse regression than the one being guarded against.

A few reasons we're keeping the targeted DELETE+INSERT as-is:
- It's the **exact same pattern** as the existing nightly export `nethcti_export.php` (`DELETE WHERE sid_imported='nethcti'` then reinsert public contacts), so it introduces no new risk relative to what already runs in production.
- The sync is **best-effort**: on failure it's logged and the request still succeeds, and the nightly export restores the state as a fallback.
- `LOCK TABLES ... WRITE` already prevents concurrent readers (Asterisk `lookup.php`, CTI customer-card) from observing a partially-updated/empty result during the operation.

A safe incremental hardening (targeted TEMPORARY-table staging built before the lock, leaving non-nethcti rows untouched) was considered but deemed not worth the added complexity for a best-effort path with a nightly fallback.

File: store/phonebook.go [hunk 2]
```diff
@@ -424,6 +424,67 @@ func DeletePhonebookEntryByID(ctx context.Context, id int64) error {
...
+			owner_id, 'nethcti', homeemail, workemail, homephone, workphone, cellphone, fax,
+			title, company, notes, name, homestreet, homepob, homecity, homeprovince,
+			homepostalcode, homecountry, workstreet, workpob, workcity, workprovince,
+			workpostalcode, workcountry, url, 'nethcti'
+		FROM cti_phonebook
+		WHERE type = 'public'
+	`
+	if _, err := conn.ExecContext(ctx, insertSelect); err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (tonyco97) [reply store/phonebook.go:501]: The production `phonebook.phonebook` schema (ns8 `40_phonebook.phonebook-schema.sql`) does include `sid_imported`, and the authoritative nightly job `nethcti_export.php` keys its cleanup on `DELETE ... WHERE sid_imported='nethcti'` while inserting rows marked with **both** `type='nethcti'` and `sid_imported='nethcti'`.

Keeping `sid_imported` here is deliberate: it makes this function operate on the exact same row set as the nightly export (full lockstep, no duplicate rows). Switching to `type` would diverge from the nightly DELETE key and leave duplicate `type='nethcti'` rows between a nightly run and the next CTI mutation.

The real gap your comment surfaced is that the **test fixture** omitted `sid_imported` — fixed by aligning `ensureCentralizedPhonebookTable` to the production schema (plus an idempotent `ADD COLUMN IF NOT EXISTS`). The query is unchanged.

File: store/phonebook.go [hunk 3]
```diff
@@ -424,6 +424,67 @@ func DeletePhonebookEntryByID(ctx context.Context, id int64) error {
 	return err
 }
 
+// SyncPublicContactsToCentralized republishes every public CTI contact into the
+// centralized phonebook used for call-time name resolution (Asterisk lookup.php and
+// the CTI customer-card "Identity" query both read phonebook.phonebook).
+//
+// It mirrors exactly the nightly nethcti_export.php job: it deletes the rows previously
+// exported by NethCTI (sid_imported = 'nethcti') and reinserts every cti_phonebook row
```
[the remark this answers is not in the document]
Comment (tonyco97) [reply store/phonebook.go]: Updated the doc comment to be explicit: NethCTI-exported rows are marked with both `type='nethcti'` and `sid_imported='nethcti'` (as `nethcti_export.php` does); the middleware union search excludes them via `type != 'nethcti'`, while this function keys cleanup on `sid_imported='nethcti'` to stay in lockstep with the nightly export. Rows from other sync sources (`sid_imported != 'nethcti'`) are left untouched.

File: store/phonebook.go [hunk 4]
```diff
@@ -424,6 +424,67 @@ func DeletePhonebookEntryByID(ctx context.Context, id int64) error {
...
+//
+// phonebook.phonebook is a MyISAM table (no transactions), so the delete+reinsert is
+// wrapped in LOCK TABLES ... WRITE to avoid exposing an empty result set to concurrent
+// readers. LOCK TABLES is connection-scoped, hence the work runs on a single pinned
+// connection.
+func SyncPublicContactsToCentralized(ctx context.Context) error {
+	database := db.GetDB()
+	if database == nil {
+		return errors.New("database not initialized")
+	}
```
[the remark this answers is not in the document]
Comment (tonyco97) [reply store/phonebook.go:461]: Added a store integration test `TestSyncPublicContactsToCentralized` that seeds `cti_phonebook` with a public and a private contact plus an external centralized row (`sid_imported='mssql'`), then asserts: the public contact is exported once with `type='nethcti'`/`sid_imported='nethcti'`, the private contact is not exported, the external row is preserved (the regression guard for the partial-update concern), and a second run is idempotent.

File: store/phonebook.go [hunk 5]
```diff
@@ -424,6 +424,70 @@ func DeletePhonebookEntryByID(ctx context.Context, id int64) error {
...
+		return err
+	}
+	defer conn.Close()
+
+	if _, err := conn.ExecContext(ctx, "LOCK TABLES `phonebook`.`phonebook` WRITE, `cti_phonebook` READ"); err != nil {
+		return err
+	}
+	// Always release the lock, even on error.
+	defer func() {
+		if _, unlockErr := conn.ExecContext(ctx, "UNLOCK TABLES"); unlockErr != nil {
```
Comment (tommaso-ascani) [inline store/phonebook.go]: **[HIGH] UNLOCK uses the expired ctx → lock leak into the pool.** The `defer` reuses the same `ctx` with the 30s deadline. If DELETE/INSERT exhaust the timeout (large table or contention on the WRITE lock), `conn.ExecContext(ctx, "UNLOCK TABLES")` returns `context.DeadlineExceeded` immediately **without sending UNLOCK** (the driver checks `ctx.Err()` before executing). Then `conn.Close()` returns a session still in `LOCK TABLES` to the pool: the next borrower inherits the lock → MySQL error 1100 `Table ... was not locked` on queries touching other tables, poisoning up to ~25 connections until `ConnMaxLifetime` recycles them.

Fix: UNLOCK with `context.Background()` (its own timeout); on UNLOCK failure discard the connection instead of returning it to the pool.

File: store/phonebook.go [hunk 6]
```diff
@@ -424,6 +424,70 @@ func DeletePhonebookEntryByID(ctx context.Context, id int64) error {
...
+			logs.Log("[ERROR][PHONEBOOK] Failed to unlock tables after centralized sync: " + unlockErr.Error())
+		}
+	}()
+
+	if _, err := conn.ExecContext(ctx, "DELETE FROM `phonebook`.`phonebook` WHERE sid_imported = 'nethcti'"); err != nil {
```
Comment (tommaso-ascani) [inline store/phonebook.go:486]: **[HIGH] Non-transactional DELETE→INSERT on MyISAM → wipe of public contacts.** MyISAM has no rollback. The DELETE commits immediately; if the following `INSERT...SELECT` times out/errors (see the UNLOCK comment) or the connection drops, the rows stay deleted. `LOCK TABLES` only hides the empty intermediate state from readers *during* the call, it does not restore. Result: every public NethCTI contact disappears from the Asterisk lookup until the next successful sync or the nightly export. This is a new failure mode: the nightly export this 'mirrors' is not request-bounded.

Fix: build into a temp table + atomic `RENAME TABLE`, or go incremental.

File: methods/phonebook_legacy.go [hunk 7]
```diff
@@ -219,6 +220,22 @@ func GetCentralizedPhonebookContact(c *gin.Context) {
...
 }
 
+// syncCentralizedPublicContacts republishes public CTI contacts into the centralized
+// phonebook so they become immediately resolvable at call time. It is best-effort: a
+// failure is logged but never fails the CRUD request (the cti_phonebook write already
+// succeeded and the nightly export remains a fallback).
+//
+// It runs with its own context/timeout, independent of the originating HTTP request, so
+// that a near-deadline (or already-cancelled) request context cannot starve the sync.
+func syncCentralizedPublicContacts() {
```
Comment (tommaso-ascani) [inline methods/phonebook_legacy.go:265]: **[HIGH] Synchronous sync in the request path → pool exhaustion.** `syncCentralizedPublicContacts()` runs *before* `c.Status()`, synchronously, pinning a connection + global WRITE lock for up to 30s. The comment says 'independent of the originating HTTP request', but only the *context* is independent, not the execution: it still blocks the response. Under a burst of public edits, ~25 waiters saturate `MaxOpenConns=25` → app-wide DB starvation (login, CDR, search). The WRITE lock also blocks `lookup.php`/customer-card readers during each republish.

Fix: run the republish off-request (goroutine/debounced queue).

File: store/phonebook.go [hunk 8]
```diff
@@ -424,6 +424,70 @@ func DeletePhonebookEntryByID(ctx context.Context, id int64) error {
...
+		return err
+	}
+
+	insertSelect := `
```
Comment (tommaso-ascani) [inline store/phonebook.go]: **[MEDIUM] O(N) republish per single-row change.** Every CRUD on a single public contact re-deletes all `sid_imported='nethcti'` rows and re-inserts *all* `type='public'` rows. A bulk of N edits → O(N²) DB work. No incremental path, no change detection. Amplifies the wipe window and pool pressure. Consider an incremental republish (only the changed row) or at least debounce.

**Review by tommaso-ascani — COMMENTED** [review]
Automated review (recall-biased, high effort). The happy-path logic is correct and the tests are good, but `SyncPublicContactsToCentralized` has three serious problems around lock/connection handling and cost, plus a few minor ones. The three main ones (#1 lock leak, #2 non-transactional wipe, #3 synchronous sync in the request path) are solved by moving the republish off the request path (async/debounced) and using `context.Background()` for UNLOCK + discarding the connection on failure.

File: store/phonebook.go [hunk 9]
```diff
@@ -424,6 +424,70 @@ func DeletePhonebookEntryByID(ctx context.Context, id int64) error {
...
+		return errors.New("database not initialized")
+	}
+
+	conn, err := database.Conn(ctx)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	if _, err := conn.ExecContext(ctx, "LOCK TABLES `phonebook`.`phonebook` WRITE, `cti_phonebook` READ"); err != nil {
```
Comment (tommaso-ascani) [inline store/phonebook.go:469]: **[MEDIUM] DELETE keys on `sid_imported` but reads filter on `type`.** Cleanup keys on `sid_imported='nethcti'`, while reads/union exclude `type != 'nethcti'`. Today the INSERT sets both → consistent. But pre-existing nethcti rows with `sid_imported` NULL/other (e.g. an older export) survive the DELETE → duplicates in the lookup. The invariant depends on lockstep with `nethcti_export.php` (another repo) and is not enforced at the schema level. Better to document/enforce it, or also delete by `type='nethcti'`.

File: methods/phonebook.go [hunk 10]
```diff
@@ -220,6 +231,10 @@ func AdminImportPhonebookCSV(c *gin.Context) {
 	response.ImportedRows = successful
 	response.FailedRows = failed
 
+	if successful > 0 && entriesContainPublic(entries) {
```
Comment (tommaso-ascani) [inline methods/phonebook.go:234]: **[LOW] `entriesContainPublic` also evaluates failed rows.** `successful > 0 && entriesContainPublic(entries)` scans *all* parsed rows, including ones that failed to import. A CSV with only a failed public row + a successful private row → a pointless full republish under the global lock. Check the actually-imported subset instead.

File: methods/phonebook_legacy.go [hunk 11]
```diff
@@ -219,6 +220,22 @@ func GetCentralizedPhonebookContact(c *gin.Context) {
...
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	if err := syncCentralizedPublicContactsFunc(ctx); err != nil {
```
Comment (tommaso-ascani) [inline methods/phonebook_legacy.go:269]: **[LOW] Sync error swallowed while the response is 200/201.** Best-effort by design, but combined with the non-transactional wipe it hides the problem: the user gets 201/200 while the centralized table is broken/empty. No metric/signal distinguishes 'cti OK, republish failed'. At least emit a distinguishable log/metric for alerting.

File: methods/phonebook_legacy.go [hunk 12]
```diff
@@ -361,6 +382,10 @@ func UpdateLegacyCTIPhonebookContact(c *gin.Context) {
 		return
 	}
 
+	if existingContact.Type == "public" || nextType == "public" {
```
Comment (tommaso-ascani) [inline methods/phonebook_legacy.go]: **[LOW] `== "public"` match vs the CSV path's `EqualFold`.** The legacy handlers use exact equality; the CSV path (`methods/phonebook.go`) uses `EqualFold`+`TrimSpace`. If a stored `Type` is not normalized (`Public`, ` public` from another writer), update/delete skips the sync → stale centralized table. Use the same comparison across all paths.

File: methods/phonebook_legacy.go [hunk 13]
```diff
@@ -284,6 +301,10 @@ func CreateLegacyCTIPhonebookContact(c *gin.Context) {
 		return
 	}
 
+	if contactType == "public" {
```
Comment (tommaso-ascani) [inline methods/phonebook_legacy.go]: **[LOW] Sync trigger duplicated across 4-5 handlers** instead of living in the store layer (`Create/Update/DeletePhonebookEntry`). Each call site must remember the right predicate (update needs `existingContact.Type=="public" || nextType=="public"`). A future sixth write path that forgets the guard silently diverges until the nightly export. Centralizing it in the store closes the gap.

File: store/phonebook.go [hunk 14]
```diff
@@ -424,6 +424,70 @@ func DeletePhonebookEntryByID(ctx context.Context, id int64) error {
...
+			logs.Log("[ERROR][PHONEBOOK] Failed to unlock tables after centralized sync: " + unlockErr.Error())
+		}
+	}()
+
+	if _, err := conn.ExecContext(ctx, "DELETE FROM `phonebook`.`phonebook` WHERE sid_imported = 'nethcti'"); err != nil {
+		return err
+	}
+
+	insertSelect := `
+		INSERT INTO ` + "`phonebook`.`phonebook`" + ` (
```
Comment (tommaso-ascani) [inline store/phonebook.go]: **[LOW] 26-column list duplicated** from `BatchInsertPhonebookEntries`. A schema change must be hand-applied in two places; a mismatch between the INSERT column list and the SELECT projection silently writes data into the wrong column. Extract a shared `[]string` constant as the single source of truth.

File: store/phonebook.go (hunk 5 above)
Comment (tonyco97) [reply store/phonebook.go]: Fixed in `a2b0a65`: UNLOCK now runs with a dedicated `context.Background()` + 5s timeout inside its own `defer`, so an expired request context cannot prevent the UNLOCK from being sent. On UNLOCK failure the code logs `[CRITICAL]` and lets `conn.Close()` return the connection; with Background() the failure window is negligible in practice.

File: methods/phonebook_legacy.go (hunk 7 above)
Comment (tonyco97) [reply methods/phonebook_legacy.go:265]: Fixed in `a2b0a65`: the republish is now fully off-request. `scheduleSyncCentralizedPublicContactsFunc()` sends to a capacity-1 channel; a single background goroutine drains it. A burst of N public edits enqueues at most two full syncs (one running, one coalesced), never N. The HTTP response is no longer blocked.

File: store/phonebook.go (hunk 6 above)
Comment (tonyco97) [reply store/phonebook.go:486]: WONTFIX — and the proposed fix (temp table + \`RENAME TABLE\`) would introduce a worse regression here.

\`phonebook.phonebook\` is **shared** across multiple import sources (MSSQL, CSV, Infinity, NethCTI). A \`RENAME TABLE\` would swap the entire table, wiping all rows from those other sources. The approach would break the system far more than the narrow wipe window it is trying to close.

The current DELETE+INSERT touches **only** \`sid_imported='nethcti'\` rows — exactly what \`nethcti_export.php\` does in the nightly job already running in production (see \`ns8-nethvoice/freepbx/usr/share/phonebooks/nethcti_export.php\` L28+L69). No new risk profile compared to the nightly. The failure window (empty nethcti rows between DELETE and INSERT) is now also narrowed by the off-request execution and \`context.Background()\` context, reducing the chance of a mid-sync timeout.

File: store/phonebook.go (hunk 8 above)
Comment (tonyco97) [reply store/phonebook.go]: Mitigated by the coalescing worker introduced in \`a2b0a65\`: a burst of N public edits fires at most two full syncs (one already running, one enqueued), never N. An incremental per-row path would be more efficient but diverges from the "mirror the nightly" model and adds significant complexity. Leaving as-is for now; can be revisited if profiling shows it is a bottleneck.

File: store/phonebook.go (hunk 9 above)
Comment (tonyco97) [reply store/phonebook.go:469]: The discriminator mismatch is pre-existing and intentional: it mirrors exactly how `nethcti_export.php` works (DELETE by `sid_imported`, INSERT with both `type` and `sid_imported` set to `"nethcti"`). The invariant — that every nethcti-exported row always carries both columns set to `"nethcti"` — is enforced by the INSERT in both the nightly and our sync, not at the schema level. Pre-existing rows with `sid_imported=NULL` cannot exist because the nightly has always written both columns. The constraint is now explicitly documented in the `SyncPublicContactsToCentralized` doc-comment.

File: methods/phonebook.go (hunk 10 above)
Comment (tonyco97) [reply methods/phonebook.go:234]: Acknowledged, but fixing it cleanly would require changing the return signature of `batchInsertPhonebookEntries` to expose which entries succeeded — a change that also touches `CreatePhonebookEntry` and the test harness. Given that the sync is now off-request and coalesced, a spurious republish in this edge case (failed public row + successful private row) costs one background job rather than blocking a request. Leaving this as a known minor inefficiency for now.

File: methods/phonebook_legacy.go (hunk 11 above)
Comment (tonyco97) [reply methods/phonebook_legacy.go:269]: Fixed in `9c0bd27`: the log level is now `[CRITICAL]` with a message that explicitly states "CTI write already committed; centralized lookup stale until next sync or nightly export". This makes the divergence visible to monitoring without changing the best-effort semantics.

File: methods/phonebook_legacy.go (hunk 12 above)
Comment (tonyco97) [reply methods/phonebook_legacy.go]: Fixed in `a2b0a65`: all legacy handlers now use `isPublicContactType()` which wraps `strings.EqualFold` + `strings.TrimSpace`, identical to the CSV path. No more `== "public"` direct comparisons.

File: methods/phonebook_legacy.go (hunk 13 above)
Comment (tonyco97) [reply methods/phonebook_legacy.go]: Agreed on the long-term risk, but centralizing the trigger inside the store layer would couple store to a background scheduling concern and could emit unintended syncs from existing callers (e.g. batch import already does its own guard). Keeping the trigger in the handler layer is explicit and auditable. Will address in a dedicated refactor PR once this feature is stable in production.

File: store/phonebook.go (hunk 14 above)
Comment (tonyco97) [reply store/phonebook.go]: Fixed in `9c0bd27`: extracted `centralizedSyncColumns` (the 24 passthrough columns, excluding `type` and `sid_imported` which are written as literal `"nethcti"` markers). The `INSERT` column clause and the `SELECT` projection are now both built from this single slice, so a schema change is applied once and column-order alignment is guaranteed by construction.

**Review by tommaso-ascani — APPROVED** [review]
